### PR TITLE
Fix minor parsing issues

### LINKED ISSUE #406: IndexOutOfRangeException when parsing `#`
I realize that this issue is not too bad, since it's technically not going to parse anyway. But if you try to parser a string that is just `#`, it throws this exception:
```
System.IndexOutOfRangeException: Index was outside the bounds of the array.
   at System.String.get_Chars(Int32 index)
   at Esprima.Scanner.ScanComments()
   at Esprima.JavaScriptParser.CollectComments()
   at Esprima.JavaScriptParser.NextToken(Boolean allowIdentifierEscape)
   at Esprima.JavaScriptParser.Reset(String code, String source)
   at Esprima.JavaScriptParser.ParseModule(String code, String source)
```
image

esprima, more correctly, returns: `Line 1: Unexpected token ILLEGAL`. 
image

### LINKED ISSUE #408: Error parsing literal with weird characters
```js
var test = {
    test: "Tests ",
}
```

Leads to:
```
Esprima.ParserException: Line 2: Unexpected token ILLEGAL
   at Esprima.Scanner.ThrowUnexpectedToken(String message)
   at Esprima.Scanner.ScanStringLiteral(Boolean strict)
   at Esprima.Scanner.Lex(LexOptions& options)
   at Esprima.JavaScriptParser.NextToken(Boolean allowIdentifierEscape)
   at Esprima.JavaScriptParser.ParseObjectProperty(Int32& protoCount)
   at Esprima.JavaScriptParser.ParseObjectInitializer()
   at Esprima.JavaScriptParser.ParsePrimaryExpression()
   at Esprima.JsxParser.ParsePrimaryExpression()
   at Esprima.JavaScriptParser.ParseLeftHandSideExpressionAllowCall()
   at Esprima.JavaScriptParser.ParseUpdateExpression()
   at Esprima.JavaScriptParser.ParseUnaryExpression()
   at Esprima.JavaScriptParser.ParseBinaryExpressionOperand()
   at Esprima.JavaScriptParser.ParseBinaryExpression()
   at Esprima.JavaScriptParser.ParseAssignmentExpression()
   at Esprima.JavaScriptParser.ParseObjectProperty(Int32& protoCount)
   at Esprima.JavaScriptParser.ParseObjectInitializer()
   at Esprima.JavaScriptParser.ParsePrimaryExpression()
   at Esprima.JsxParser.ParsePrimaryExpression()
   at Esprima.JavaScriptParser.ParseLeftHandSideExpressionAllowCall()
   at Esprima.JavaScriptParser.ParseUpdateExpression()
   at Esprima.JavaScriptParser.ParseUnaryExpression()
   at Esprima.JavaScriptParser.ParseBinaryExpressionOperand()
   at Esprima.JavaScriptParser.ParseBinaryExpression()
   at Esprima.JavaScriptParser.ParseAssignmentExpression()
   at Esprima.JavaScriptParser.IsolateCoverGrammar[T](Func`1 parseFunction)
   at Esprima.JavaScriptParser.ParseVariableDeclaration(Boolean inFor)
   at Esprima.JavaScriptParser.ParseVariableDeclarationList(Boolean inFor)
   at Esprima.JavaScriptParser.ParseVariableStatement()
   at Esprima.JavaScriptParser.ParseStatement()
   at Esprima.JavaScriptParser.ParseStatementListItem()
   at Esprima.JavaScriptParser.ParseModule(String code, String source)
```

The same code is parsed by acorn, and is accepted by Chrome:
image

### PR DESCRIPTION
**Pull request by adams85**
Fixes #406 
Fixes #408 

Also fixes some minor issues in the related areas (see commit messages).

### CODE REVIEW

**Comment by CharlieEriksen** [comment]
Thanks for the quick fixes @adams85!

While looking over the CLs, it reminded me of something I've been trying to figure out the last few days:

There's no longer a `Strict` flag on the parser options. But there are still references to `strict` in the code. What happened to the flag?

File: src/Esprima/Esprima.csproj [hunk 1]
```diff
@@ -1,6 +1,6 @@
 ﻿<Project Sdk="Microsoft.NET.Sdk">
   <PropertyGroup>
-    <Description>Standard-compliant ECMAScript 2019 parser (also known as JavaScript).</Description>
+    <Description>Standard-compliant ECMAScript 2022 parser (also known as JavaScript).</Description>
```
Comment (lahma) [inline src/Esprima/Esprima.csproj:3]: 👍🏻

**Review by lahma — APPROVED** [review]
LGTM!

**Comment by adams85** [comment]
@lahma
Thanks for the review!

@CharlieEriksen
> There's no longer a Strict flag on the parser options. But there are still references to strict in the code. What happened to the flag?

The parser was refactored to be reusable (setup one instance, parse multiple scripts with the same settings) The strict setting was removed from the options so you don't need to create multiple parser instances when you want to parse a script in non-strict mode and another one in strict mode. Now you can [pass this flag to the `ParseScript` method](https://github.com/sebastienros/esprima-dotnet/blob/v3.0.0/src/Esprima/JavaScriptParser.cs#L247) directly.

In case you wonder why you don't have this option in the case of `ParseModule`, that's because strict mode is enabled in ES modules by default.

And if you wonder why you don't have this option in the case of `ParseExpression`, erm.. I just realized that we forgot to add the strict mode option to that method. Gonna fix this soon, too.

**Pull request merged**